File: src/test/java/com/uapp_llc/test/model/factory/FactoryProducer.java
package com.uapp_llc.test.model.factory;

import java.util.HashMap;
import java.util.Map;

import com.uapp_llc.model.Column;
import com.uapp_llc.model.Task;

final class FactoryProducer {

  private static final Map<String, AbstractFactory<?>> typeFactories = new HashMap<>();

  private FactoryProducer() {
  }

  static <T> AbstractFactory<T> getFactory(Class<T> type) {
    String typeName = type.getName();

    if (!typeFactories.containsKey(typeName)) {
      if (Column.class.equals(type)) {
        typeFactories.put(typeName, new ColumnFactory());
      }

      if (Task.class.equals(type)) {
        typeFactories.put(typeName, new TaskFactory());
      }
    }

    return (AbstractFactory<T>) typeFactories.get(typeName);
  }

}

File: src/test/kotlin/com/uapp_llc/test/model/factory/FactoryProducer.kt
package com.uapp_llc.test.model.factory

import com.uapp_llc.model.Column
import com.uapp_llc.model.Task

object FactoryProducer {

  private val typeFactories: MutableMap<String, AbstractFactory<*>> = mutableMapOf()

  @JvmStatic
  fun <T> getFactory(type: Class<T>): AbstractFactory<T> {
    val typeName: String = type.typeName

    if (!typeFactories.containsKey(typeName)) {
      when (type) {
        Column::class.java -> typeFactories[typeName] = ColumnFactory()
        Task::class.java -> typeFactories[typeName] = TaskFactory()
      }
    }

    return typeFactories[typeName] as AbstractFactory<T>
  }

}

